JOSESwift/Sources/JWEHeader.swift

```diff
@@ -25,8 +25,14 @@ import Foundation
 
 /// The header of a `JWE` object.
 public struct JWEHeader: JOSEHeader {
-    let headerData: Data
-    let parameters: [String: Any]
+    var headerData: Data
+    var parameters: [String: Any] {
+        didSet {
+            // Forcing the try is ok here, since [String: String] can be converted to JSON.
+            // swiftlint:disable:next force_try
+            headerData = try! JSONSerialization.data(withJSONObject: parameters, options: [])
```

Comment:
I am not sure when to exactly update the `headerData`. I currently use a `didSet` on the `parameter` property, so the `headerData` gets updated every time a parameter is set. Is this a solution which works for you?

Reply:
I think this is perfectly fine 👌 😍

Reply:
Hi @fbernutz, 

that's a perfect place to do the data computation.

Just one thing: a `[String: Any]` dictionary that contains an URL won't be JSON encodable. The `try!` will fail here if the `parameters` contain a URL. What would you think about converting the URLs to `String` before setting them in the parameters dictionary and converting the `String` back to a `URL` when getting the parameter. E.g. like this:

``` swift
public var jku: URL? {
  set {
    parameters["jku"] = newValue.absoluteString
  }
  get {
    return URL(string: parameters["jku"])
   }
}	    
```

Reply:
Hi @daniel-mohemian 🙂 
What should happen if it is somehow not a valid JSON? I thought about throwing an error like this: `HeaderParsingError.headerIsNotValidJSONObject`. But who could catch the thrown error? I am not quite sure from where the `headerData` is read. I could just return and don't update the headerData. What do you think about it?

Reply:
> What should happen if it is somehow not a valid JSON?

Good point. We should never get to this point (as `parameters` is internal and your setters guarantee either `String` or `[String]`). The `headerData` is only used for serializing, and most importantly verifying a JWS / decrypting a JWE. We need to keep this `headerData` separate from the parameters dictionary because a Swift dictionary does not guarantee fixed ordering of its values. However, for e.g. signature validation it is crucial that we verify against the header parameters in the order we received them.

I admit, our current implementation is a bit awkward. But for now, we should be fine. As long as `JSONSerialization.isValidJSONObject(parameters)` is true (which is currently guaranteed by your implementation) we will not get in an inconsistent sate where `headerData` and `parameter` differ. We hope to refactor those things in the future.